packages/charts/src/components/legend/legend.tsx

```diff
@@ -0,0 +1,203 @@
+        return serie.color[0]
+      })
+    }, [series])
+
+    /**
+     * Pure function that changes the state of the legend.
+     * Based on the type of action and the index of the change,
+     * the legend can decide what is the next state to be rendered.
+     *
+     * @param {number} index - Index of the serie to change the state.
+     * @param {string} type - Type of action to perform. Can be 'toggle', 'selectAll' or 'exclusive'.
+     * @returns {LegendState} The new state of the legend.
+     */
+    const changeState = useCallback(
+      (index: number, type: LegendAction['type']) => {
+        const newState = [...seriesState] as LegendState
+        if (type === 'selectAll') {
+          newState.forEach((serie) => {
+            serie.state = 'unchecked'
+          })
+        } else if (type === 'exclusive') {
+          newState.forEach((serie, i) => {
+            if (index !== i) serie.state = 'off'
+            else serie.state = 'checked'
+          })
+        } else {
+          if (index >= seriesState.length)
+            return setSeriesState(checkAllSelected(newState))
+          newState[index].state =
+            newState[index].state !== 'off' ? 'off' : 'checked'
+        }
+        const checkedState = checkAllSelected(newState)
+        setSeriesState(checkedState)
+        return checkedState
+      },
+      [seriesState]
+    )
+
+    useImperativeHandle(
+      ref,
+      () => ({
+        setState: (index: number, action: string) => {
+          console.log(action)
+          changeState(index, action as LegendAction['type'])
+        },
+      }),
+      [changeState]
+    )
+
+    const onClick = useCallback(
+      (name: string) => {
+        if (!chartRef.current) return
+        const chart = chartRef.current.getEchartsInstance()
+        const index = seriesState.findIndex((serie) => serie.serie === name)
+        const newState = [...seriesState]
+
+        const on: string[] = []
+        const off: string[] = []
+        const checked: string[] = []
+        newState.forEach((serie) => {
+          if (serie.state === 'checked') on.push(serie.serie)
+          else if (serie.state === 'off') off.push(serie.serie)
+          else checked.push(serie.serie)
+        })
+
+        let action = ''
+        if (checked.length !== 0) {
+          action = 'exclusive'
+        } else if (on.length === 1 && on[0] === name) {
+          action = 'selectAll'
+        } else {
+          action = 'toggle'
+        }
+
+        const settedState = changeState(index, action as LegendAction['type'])
+        chart.dispatchAction({
+          type: 'legendToggleSelect',
+          // repurposing this field for the extra info we need
+          name: {
+            index: index,
+            type: action,
+            state: settedState,
+            chartId: chart.getId(),
+          },
+        })
+      },
+      [seriesState, chartRef, changeState]
+    )
```

Comment:
Some tips: use meaningful names instead of "on",  "off",  etc.; avoid magic numbers; create constants like enum instead of hardcoded strings like `exclusive`, `selectAll`, etc.